compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -512,6 +512,13 @@ pub struct CopyNonOverlapping<'tcx> {
 ///     must also be `cleanup`. This is a part of the type system and checked statically, so it is
 ///     still an error to have such an edge in the CFG even if it's known that it won't be taken at
 ///     runtime.
+///  4. The induced subgraph on cleanup blocks must look roughly like an upside down tree. This is
+///     necessary to ensure that landing pad information can be correctly codegened. More precisely:
```

Comment:
What is the "subgraph on cleanup blocks"? Is it the graph where you remove all non-cleanup blocks and only keep edges between cleanup blocks?

What exactly is forbidden here? Is it essentially any kind of merging control flow? As in, every cleanup block must only have one cleanup block predecessor -- is that the condition? The analysis you implemented is a lot more complicated than that, and I am lost in all the domination talk...

Reply:
> What is the "subgraph on cleanup blocks"? Is it the graph where you remove all non-cleanup blocks and only keep edges between cleanup blocks?

Yes. I'll spell this out. I also realize this should have said "subgraph on *reachable* cleanup blocks" so will fix that as well.

> Or does the "upside down" mean that every cleanup block most only have one cleanup block successor?

This is the idea, but only after contracting dominators. One of the things that makes this constraint hard to understand is that there are no local structures that are disallowed in the graph. Indeed, if the cleanup blocks have just one entrance point, then there are no restrictions at all (since that entrance point will necessarily dominate all other reachable cleanup blocks). We take some advantage of this too - drop elab actually generates cycles (for dropping arrays) and branches (for drop flags) .

Instead this constraint is about how control flow can merge "between entrance points to cleanup code." One simple example of a graph that is disallowed is the one above, another example is a W shaped graph (all edges pointing down).

Reply:
Ah, I was hoping I could avoid thinking about dominators. ;) I never delved deeply enough into compilers to get solid intuition for them... In this case, do you compute dominators before 

So, an example of something we cannot have is a control flow split where one of the successors is also reachable from another independent cleanup path? But a "V", where two independent cleanup paths merge, is okay?

Reply:
An alternative definition that doesn't use dominators. 

It must be possible to partition reachable cleanup blocks into groups such that:

* Each group has a unique entry block. Edges from outside into the group must target the group's entry block.
* Edges from within a particular group to outside must target the same block.
* Edges in-between groups cannot form a cycle.
* There are no restriction on edges within the group.

Reply:
> I never delved deeply enough into compilers to get solid intuition for them... In this case, do you compute dominators before

End of the sentence may have gotten lost there? Otherwise you'll have to expand a bit

> So, an example of something we cannot have is a control flow split where one of the successors is also reachable from another independent cleanup path? But a "V", where two independent cleanup paths merge, is okay?

Yes.

Reply:
> Edges from outside into the group must target the group's entry block.

Obviously a nit but note that we have to restrict to edges from reachable blocks

Reply:
> End of the sentence may have gotten lost there? Otherwise you'll have to expand a bit

Sorry... got interrupted and then forgot to finish the sentence.^^

I was about to ask whether dominance is computed before or after restricting the graph to cleanup blocks.

Reply:
> I was about to ask whether dominance is computed before or after restricting the graph to cleanup blocks. The comment should clarify that.

Dominance is traditionally only defined for a control flow graph (ie a directed graph with a "start" vertex), so it doesn't really make sense to compute it after restricting to cleanup blocks, which typically do not form a control flow graph. That being said, there is a fairly natural extension of the definition (in which we consider a set of start vertices, instead of just one). In that case it doesn't matter which we pick. I'll clarify anyway.

> For me personally this restriction is so arbitrary and unmotivated that the examples would also be helpful, but it is very possible that this makes more sense to others. :)

Yeah, I don't *really* understand it either, but I also don't feel like going and reading the Itanium ABI spec to find out...

Reply:
The constraints are described in https://llvm.org/docs/ExceptionHandling.html#funclet-transitions. Their relation to MIR can only be understood in the context of code generation for MSVC target (other targets don't use funclets).

Reply:
Oh I thought Itanium did as well. Interesting